pkg/prometheus/promcfg.go

```diff
@@ -811,8 +811,51 @@ func (cg *configGenerator) generateProbeConfig(
 	return cfg
 }
 
+func mountTLSConfigFromSecretOrConfigmap(tls *v1.TLSConfig, promSpec *v1.PrometheusSpec) {
+
+	// it is guaranteed that if KeySecret is set then Cert (SecretOrConfigMap) is set as well.
+	// Cert can be a Secret (either the same as KeySecret or a different one), or a ConfigMap.
+	var keySecretName, certResourceName, caResourceName, certPrefixedResourceName,
+		caPrefixedResourceName string
+
+	if tls != nil {
+		if tls.KeySecret != nil {
+			keySecretName = tls.KeySecret.LocalObjectReference.Name
+			promSpec.Secrets = append(promSpec.Secrets, keySecretName)
+			if tls.Cert.Secret != nil {
+				certResourceName = tls.Cert.Secret.LocalObjectReference.Name
+				certPrefixedResourceName = "secret-" + certResourceName
+				if keySecretName != certResourceName {
+					promSpec.Secrets = append(promSpec.Secrets, certResourceName)
+				}
+			} else {
+				certResourceName = tls.Cert.ConfigMap.LocalObjectReference.Name
+				certPrefixedResourceName = "configmap-" + certResourceName
+				promSpec.ConfigMaps = append(promSpec.ConfigMaps, certResourceName)
+			}
+		}
+
+		if (tls.CA != v1.SecretOrConfigMap{}) {
+			if tls.CA.Secret != nil {
+				caResourceName = tls.CA.Secret.LocalObjectReference.Name
+				caPrefixedResourceName = "secret-" + caResourceName
+				if caResourceName != keySecretName && caPrefixedResourceName != certPrefixedResourceName {
+					promSpec.Secrets = append(promSpec.Secrets, caResourceName)
```

Comment:
if this is the strategy we want to go with, then this function should rather return volumes and volumemounts, modifying the prometheus object with extra information inferred from some other settings is not the right thing to do here .. we kind of have precedence of this type of thing, which is the alertmanager endpoint TLS configuration being in the same namespace as the prometheus, I think that strategy would be what we should do for remote write as well

Reply:
I am not sure I completely understand what you mean here, I tough that the point of `prometheus.Spec.Secrets`/'prometheus.Spec.ConfigMaps` is to describe the names of the secrets/configmaps to be mounted automatically at the StatefulSet stage. 

This function isn't mounting actually but just mention the names of the resources to be mounter later in `pkg/prometheus/statefulset.go`. Which volumes/volumemountes would you expect to be returned? 

Can you please elaborate a bit more about you preferred solution - "alertmanager endpoint TLS configuration"?

Reply:
Sorry it appears the alertmanager functionality doesn't actually exist. What I was trying to say is `prometheus.Spec.ConfigMaps` is only ever supposed to be used by users, the operator should not set them in order to force some behavior, that would be a cyclic dependency in some way. We've violated this in the early days of the Prometheus Operator a bit, but we should aim for never modifying the objects coming in to be processed, this only sets up for confusion of what modifies the object where.